src/policy/mallocspace/metadata.rs

```diff
@@ -0,0 +1,177 @@
+
+use std::collections::HashSet;
+use std::sync::RwLock;
+
+lazy_static! {
+    pub static ref ACTIVE_CHUNKS: RwLock<HashSet<Address>> = RwLock::default();
+}
+
+// We use the following hashset to assert if bits are set/unset properly in side metadata.
+#[cfg(debug_assertions)]
+const ASSERT_METADATA: bool = false;
+
+#[cfg(debug_assertions)]
+lazy_static! {
+    pub static ref ALLOC_MAP: RwLock<HashSet<ObjectReference>> = RwLock::default();
+    pub static ref MARK_MAP: RwLock<HashSet<ObjectReference>> = RwLock::default();
+}
+
+pub(super) const ALLOC_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+    scope: SideMetadataScope::PolicySpecific,
+    offset: 0,
+    log_num_of_bits: 0,
+    log_min_obj_size: constants::LOG_MIN_OBJECT_SIZE as usize,
+};
+
+pub(super) const MARKING_METADATA_SPEC: SideMetadataSpec = SideMetadataSpec {
+    scope: SideMetadataScope::PolicySpecific,
+    offset: ALLOC_METADATA_SPEC.offset
+        + meta_bytes_per_chunk(
+            ALLOC_METADATA_SPEC.log_min_obj_size,
+            ALLOC_METADATA_SPEC.log_num_of_bits,
+        ),
+    log_num_of_bits: 0,
+    log_min_obj_size: constants::LOG_MIN_OBJECT_SIZE as usize,
+};
+
+pub fn is_meta_space_mapped(address: Address) -> bool {
+    let chunk_start = conversions::chunk_align_down(address);
+    ACTIVE_CHUNKS.read().unwrap().contains(&chunk_start)
+}
+
+pub fn map_meta_space_for_chunk(chunk_start: Address) {
+    let mut active_chunks = ACTIVE_CHUNKS.write().unwrap();
+    if active_chunks.contains(&chunk_start) {
+        return;
+    }
+    active_chunks.insert(chunk_start);
+    let mmap_metadata_result = try_map_metadata_space(
+        chunk_start,
+        BYTES_IN_CHUNK,
+        0,
+        meta_bytes_per_chunk(
+            ALLOC_METADATA_SPEC.log_min_obj_size,
+            ALLOC_METADATA_SPEC.log_num_of_bits,
+        ) + meta_bytes_per_chunk(
+            MARKING_METADATA_SPEC.log_min_obj_size,
+            MARKING_METADATA_SPEC.log_num_of_bits,
+        ),
+    );
+    debug_assert!(mmap_metadata_result, "mmap sidemetadata failed");
+}
+
+// Check if a given object was allocated by malloc
+pub fn is_alloced_by_malloc(object: ObjectReference) -> bool {
+    is_meta_space_mapped(object.to_address()) && is_alloced(object)
+}
+
+pub fn is_alloced(object: ObjectReference) -> bool {
+    is_alloced_object(object.to_address())
+}
+
+pub fn is_alloced_object(address: Address) -> bool {
+    #[cfg(debug_assertions)]
+    if ASSERT_METADATA {
+        // Need to make sure we atomically access the side metadata and the map.
+        let lock = ALLOC_MAP.read().unwrap();
+        let check =
+            lock.contains(&unsafe { address.align_down(BYTES_IN_WORD).to_object_reference() });
+        let ret = load_atomic(ALLOC_METADATA_SPEC, address) == 1;
+        debug_assert_eq!(
+            check,
+            ret,
+            "is_alloced_object(): alloc bit does not match alloc map, address = {} (aligned to {}), meta_start = {}",
```

Comment:
I don't understand the goal of `meta_start` here.
This will not work in the new metadata implementation as is.

Reply:
I was just using it to know where the metadata is stored for debugging. Why would it not work for the new implementation? We still could have a way to know where the metadata is, right?

Reply:
From what I see in `meta_start(addr)`'s implementation, it returns the starting address of the metadata chunk which contains the metadata for `addr`.
This is specific to the per-chunk metadata.
In the new implementation, we only have per-chunk metadata for the policy-specific side metadata of the 32-bit targets.

If you use `address_to_meta_address()` instead of `address_to_meta_chunk_addr()` in the implementation of `meta_start(addr)`, that will work in the new implementation.

Reply:
I see. I can make that change.

Reply:
Removed `SideMetadataSpec.meta_start()`. Use `address_to_meta_address()` for those debug prints.